[builder error] generic option (error: '_option_main__ABCD' undeclared)

### Describe the bug

```v
module main

struct ObjectDesc {
	typ	u32
	ptr voidptr
}

struct ABCD {
	name string
}

pub fn cast_object_desc[H](desc &ObjectDesc) ?H {
	$if H is &ABCD {
		/*if desc.typ == 12 { // desc == ABCD
			return &ABCD(desc.ptr)
		}*/
		if desc.typ == 12 { // desc == ABCD
			return ?&ABCD(&ABCD(desc.ptr))
		}
	}
	return none
}

fn main() {
	obj := ABCD{ name: "Foo" }
	desc := ObjectDesc{ typ:12, ptr:voidptr(&obj) }
	obj2 := cast_object_desc[&ABCD](desc) or { panic("wwww") }
	//obj2 := &ABCD(desc.ptr)
	println(obj2.name)
}
```


### Expected Behavior

readable error or ok

### Current Behavior

```
C:/Users/russo/AppData/Local/Temp/v_0/bug.16014977718429259782.tmp.c:7621: warning: cast between pointer and integer of different size
C:/Users/russo/AppData/Local/Temp/v_0/bug.16014977718429259782.tmp.c:12640: error: '_option_main__ABCD' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

`v bug.v`

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 fc4c431

### Environment details (OS name and version, etc.)

V full version: V 0.3.3 https://github.com/vlang/v/commit/b2a71ecab2ba1f574344f0ccfb835e116441b1e1.fc4c431
OS: windows, Microsoft Windows 10 Pro v19042 64-bit
Processor: 16 cpus, 64bit, little endian,

getwd: C:\Users\russo\Desktop
vexe: D:_dev\v_lang\v.exe
vexe mtime: 2023-04-04 20:52:46

vroot: OK, value: D:_dev\v_lang
VMODULES: OK, value: C:\Users\russo.vmodules
VTMP: OK, value: C:\Users\russo\AppData\Local\Temp\v_0

Git version: git version 2.35.3.windows.1
Git vroot status: weekly.2023.14-9-gfc4c431d
.git/config present: true

CC version: Error: exec failed (CreateProcess) with code 2: Не удается найти указанный файл.
cmd: cc --version
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -56,9 +56,10 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 	} else {
 		stmt_str := g.go_before_stmt(0).trim_space()
 		styp := g.base_type(ret_typ)
+		decl_styp := g.typ(ret_typ).replace('*', '_ptr')
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
-		g.writeln('${g.typ(ret_typ).replace('*', '_ptr')} ${tmp_var};')
+		g.writeln('${decl_styp} ${tmp_var};')
 		g.write('_option_ok(&(${styp}[]) {')
 
 		if expr is ast.CastExpr && expr_typ.has_flag(.option) {

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -58,7 +58,7 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 		styp := g.base_type(ret_typ)
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
-		g.writeln('${g.typ(ret_typ)} ${tmp_var};')
+		g.writeln('${g.typ(ret_typ).replace('*', '_ptr')} ${tmp_var};')
 		g.write('_option_ok(&(${styp}[]) {')
 
 		if expr is ast.CastExpr && expr_typ.has_flag(.option) {

--- a/vlib/v/tests/option_ptr_cast_test.v
+++ b/vlib/v/tests/option_ptr_cast_test.v
@@ -1,0 +1,34 @@
+struct ObjectDesc {
+	typ u32
+	ptr voidptr
+}
+
+struct ABCD {
+	name string
+}
+
+pub fn cast_object_desc[H](desc &ObjectDesc) ?H {
+	$if H is &ABCD {
+		/*
+		if desc.typ == 12 { // desc == ABCD
+			return &ABCD(desc.ptr)
+		}*/
+		if desc.typ == 12 { // desc == ABCD
+			return ?&ABCD(&ABCD(desc.ptr))
+		}
+	}
+	return none
+}
+
+fn test_option_ptr() {
+	obj := ABCD{
+		name: 'Foo'
+	}
+	desc := ObjectDesc{
+		typ: 12
+		ptr: voidptr(&obj)
+	}
+	obj2 := cast_object_desc[&ABCD](desc) or { panic('wwww') }
+	// obj2 := &ABCD(desc.ptr)
+	assert obj2.name == 'Foo'
+}

